tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/annotations/AnnotationHelper.java

```diff
@@ -0,0 +1,156 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.event.metrics.annotations;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ImmutableList;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+
+public final class AnnotationHelper {
+
+    private AnnotationHelper() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Annotation as implemented on passed in type or parent of that type, works for both super classes and interfaces.
+     *
+     * @param clazz - Class type to scan for annotations
+     * @param annotation - Annotation type to scan for
+     * @return - First matching annotation found in depth first search, or null if not found
+     */
+    public static <T extends Annotation> T getSuperTypeAnnotation(Class<?> clazz, Class<T> annotation) {
+        if (clazz.isAnnotationPresent(annotation)) {
+            return clazz.getAnnotation(annotation);
+        }
+
+        for (Class<?> ifaces : getParentClasses(clazz)) {
+            T superAnnotation = getSuperTypeAnnotation(ifaces, annotation);
+            if (superAnnotation != null) {
+                return superAnnotation;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Depth first search up the Type hierarchy to find a matching annotation,  Types which do not implement the
+     * specified method signature are ignored.
+     *
+     * @param annotation - Annotation type to scan for
+     * @param clazz - Class type to scan for matching annotations
+     * @param methodSignature - Method to search annotation for
+     * @return - First found matching annotation or null
+     */
+    public static <T extends Annotation> T getMethodAnnotation(
+            Class<T> annotation, Class<?> clazz, MethodSignature methodSignature) {
+
+        Method method;
+        try {
+            method = clazz.getMethod(methodSignature.getMethodName(), methodSignature.getParameterTypes());
+        } catch (NoSuchMethodException e) {
+            return null;
+        }
+
+        if (method.isAnnotationPresent(annotation)) {
+            return method.getAnnotation(annotation);
+        }
+
+        for (Class<?> iface : getParentClasses(clazz)) {
+            T foundAnnotation = getMethodAnnotation(annotation, iface, methodSignature);
+            if (foundAnnotation != null) {
+                return foundAnnotation;
+            }
+        }
+
+        return null;
+    }
```

Comment:
```
    /**
     * Depth first search up the Type hierarchy to find a matching annotation,  Types which do not implement the
     * specified method signature are ignored.
     *
     * @param annotation annotation class
     * @param clazz containing class
     * @param method method
     * @param <T> annotation type
     * @return - First found matching annotation or null
     */
    @Nullable
    public static <T extends Annotation> T getAnnotationIfPresent(Class<T> annotation, Class<?> clazz, Method method) {
        return getAnnotationIfPresent(annotation, clazz, method, method.getParameterTypes());
    }

    /**
     * Depth first search up the Type hierarchy to find a matching annotation. Types which do not implement the
     * specified method signature are ignored.
     *
     * @param annotationClass annotation class
     * @param clazz containing class
     * @param method method
     * @param parameterTypes method parameter types
     * @param <T> annotation type
     * @return - First found matching annotation or null
     */
    @Nullable
    private static <T extends Annotation> T getAnnotationIfPresent(
            Class<T> annotationClass,
            Class<?> clazz,
            Method method,
            Class... parameterTypes) {

        if (method == null) {
            return null;
        }

        T maybeAnnotation = method.getAnnotation(annotationClass);
        if (maybeAnnotation != null) {
            return maybeAnnotation;
        }

        String methodName = method.getName();
        for (Class<?> parentClass : getAllParentClasses(clazz)) {
            for (Method parentMethod : parentClass.getMethods()) {
                if (methodName.equals(parentMethod.getName())) {
                    Class<?>[] parentMethodParamTypes = parentMethod.getParameterTypes();
                    if (Arrays.equals(parameterTypes, parentMethodParamTypes)) {
                        maybeAnnotation = parentMethod.getAnnotation(annotationClass);
                        if (maybeAnnotation != null) {
                            log.info("Found {} on matching method {} on {}",
                                    maybeAnnotation, parentMethod, parentClass);
                            return maybeAnnotation;
                        }
                    }
                }
            }
        }

        return null;
    }
```

Reply:
@schlosna , are you advocating ditching the MethodSignature as the argument here and going back to either (Method) or (String, Class[])?

Reply:
I had branched off your previous changes and was testing out some things. We probably need some test cases where method name is the same but signatures differ by argument types.